### ISSUE TITLE
`tf32_off` sets an environment variable that nothing reads, and it is the canonical worked example throughout the probe documentation

### ISSUE THREAD

**Issue by vivekkhandelwal1** [ROCm/aorta#500]
## Summary

The built-in mitigation `tf32_off` (`src/aorta/registry/mitigations.py:37`) sets
`DISABLE_TF32=1`, annotated *"consumed by hipBLASLt itself"*. Nothing consumes it.
`DISABLE_TF32` is not an environment variable read by hipBLASLt, rocBLAS, PyTorch or
any other ROCm component, on any architecture. A cell labelled `tf32_off` is
therefore byte-for-byte a second `none` cell, which is the exact failure a mitigation
axis exists to prevent.

This would be a small registry defect except that `tf32_off` is the **worked example**
in the probe documentation — `docs/probe/usage.md`, `docs/env-probe.md`, the probe-188
rubric and the buck2 reference all teach it — so a user following the docs builds a
two-cell matrix that is the same baseline twice, sees the mitigation cell recorded with
`{"DISABLE_TF32": "1"}` in its `env`, and concludes the mitigation was tried.

## The measurement

`gfx950` (MI350X, `gfx950:sramecc+:xnack-`), container `rocm/primus:v26.3`, torch
`2.10.0+git94c6e04`, HIP `7.2.53211`, host ROCm 7.0.2.2. A 4096×4096 fp32 matmul,
Frobenius relative error against a float64 reference of the same operands:

| condition | `fp32_precision` readback | rel. error |
|---|---|---|
| `allow_tf32=False` | `ieee` | 1.15e-06 |
| `allow_tf32=True` | `tf32` | 4.49e-06 |
| `allow_tf32=True` + **`DISABLE_TF32=1`** | `tf32` | **4.49e-06** |
| `allow_tf32=False` + `DISABLE_TF32=1` | `ieee` | 1.15e-06 |
| `allow_tf32=True` + `HIPBLASLT_ALLOW_TF32=0` | `tf32` | 4.49e-06 |
| `allow_tf32=True` + `TORCH_ALLOW_TF32_CUBLAS_OVERRIDE=1` | `tf32` | 4.44e-06 |

`DISABLE_TF32` changes nothing, in either direction, from either starting state.

**On the controls, because one of them is weak and it matters which.**
`HIPBLASLT_ALLOW_TF32` was removed from PyTorch upstream in September 2025, so on this
build it is itself a dead variable and its null result proves little. The control that
carries the argument is `TORCH_ALLOW_TF32_CUBLAS_OVERRIDE`, which **moved the number**
(4.44e-06 against 4.49e-06). The harness detects an environment variable having an
effect when one does; the null for `DISABLE_TF32` is a property of the variable.

**The one-line clincher.** In an archived probe matrix from a GPU run, the
`tf32_off-none` cell's *recorded* `env` — read back out of `trial_*/result.json`, not
assumed — carries `{"DISABLE_TF32": "1"}`, **and the cell fails 4/4**, identically to
the `none-none` cell beside it. The variable demonstrably reached the process. It did
nothing.

## Scoping: who actually reads `DISABLE_TF32`?

Nobody, and this is not architecture-specific. Three independent lines:

1. **Nothing in the image contains the string.** A whole-filesystem sweep of `/opt`
   and `/usr` in `rocm/primus:v26.4` (torch 2.12, ROCm 7.14) finds the literal
   `DISABLE_TF32` in no binary, shared library or Python source — not
   `libhipblaslt.so`, not `librocblas.so`, not `libtorch_hip.so`. The same sweep
   *does* find `TORCH_ALLOW_TF32_CUBLAS_OVERRIDE` in `libtorch_cpu.so`, so the method
   finds env-var strings when they are present. For completeness, the TF32-adjacent
   knob hipBLASLt does expose is `HIPBLASLT_OVERRIDE_COMPUTE_TYPE_XF32`.

2. **The name belongs to another project.** `DISABLE_TF32` is a Torch-TensorRT Python
   constant — `torch_tensorrt.dynamo._defaults.DISABLE_TF32`, the default for the
   `disable_tf32` compile option, which controls TF32 accumulation in TensorRT layers.
   It is NVIDIA/TensorRT-side and it is a module constant, never an environment
   variable. That is the most likely origin of the name.

3. **On ROCm, TF32 is not env-var-controlled at all.** TF32 *is* real on ROCm — on
   gfx942 / MI300 via hipBLASLt — but it is selected through
   `torch.backends.cuda.matmul.allow_tf32` / `fp32_precision`, an in-process API. The
   only ROCm TF32 environment variable there has ever been is `HIPBLASLT_ALLOW_TF32`;
   it *enabled* rather than disabled, and it was removed in September 2025.

⇒ Of the three possible explanations, this is the strongest one: **not "read but
ineffective on gfx950", and not "effective on some other architecture" — simply not
read.** There is no architecture on which setting an environment variable turns TF32
off in PyTorch-on-ROCm, so this cannot be reproduced away on MI300.

Two things this does **not** claim. It does not claim TF32 is unavailable on ROCm; it
is available on MI300 through the torch API. And it does not claim no image anywhere
reads `DISABLE_TF32` — a training script is free to read anything. But a variable that
only works because a workload's Python reads it is, by this repo's own rule, not a
built-in (see below).

Separately, and not needed for the above: on gfx950 the in-process knob is barely a
precision axis either. Asked for `tf32` the stack costs about 5× IEEE, where a
software 10-bit-mantissa reference — what TF32 *means* — costs 7.7e-04, three orders
of magnitude more. Whatever gfx950 does when asked for `tf32` in this build, it is not
a 10-bit mantissa.

## Why this matters: the docs teach it

`tf32_off` is not an incidental registry entry, it is the example:

- `docs/probe/usage.md:60` — the example artifact directory is `tf32_off-none/`
- `docs/probe/usage.md:81` — the example `mitigation_axis` is `[none, tf32_off]`
- `docs/probe/usage.md:193` — the example artifact path is `tf32_off-none/`
- `docs/probe/usage.md:205` — listed among the registered names a user may use
- `docs/env-probe.md:1499` — attributed to library `pytorch`, category `gemm_numeric`
- `docs/plans/aorta-probe-188-rubric.md:68` — leads the example axis
- `docs/buck2-build-reference.md:212,226` — `tf32_off  aorta  DISABLE_TF32=1`
- `src/aorta/probe/recipe_builder.py:166` — docstring: "``tf32_off`` flips ``DISABLE_TF32``"
- `src/aorta/run/dispatcher.py:1200` — comment uses `DISABLE_TF32=1` as the example
- `recipes/llm-determinism/example-llm-determinism.yaml:62` — a committed four-cell
  recipe whose only non-`none` cell is `tf32_off`, described as depending on "TF32
  reduction ordering"

The user-visible failure mode is quiet and convincing. The run completes, the matrix
renders two differently-named cells, the mitigation cell's recorded `env` shows
`DISABLE_TF32=1` — every check aorta offers for "did the mitigation reach the
workload?" passes — and the two cells agree because they are the same experiment. If
the baseline fails, the honest conclusion "`tf32_off` did not help" is indistinguishable
from "`tf32_off` was never applied".

## The repo already disagrees with itself about this

Worth noting because it is cheap to verify and it is the shortest route to consensus:

- `src/aorta/registry/mitigations.py:37` says hipBLASLt consumes it.
- `src/aorta/registry/README.md:34` uses "`DISABLE_TF32` (hipBLASLt reads it)" as the
  canonical **qualifying** example of a built-in.
- `src/aorta/instrumentation/env_knobs.py:183` says `library="pytorch"`, consumer
  "disables the TF32/xf32 compute path" — a *different* consumer for the same
  variable — and marks it `source_reference=INHERITED_UNAUDITED`, which the file
  itself defines as *"inherited from schema <= 1.14 (PR #306 / #308); classification
  not traced to an upstream call site"*.

So one file already records that this attribution was never traced to a call site,
while another presents it as the textbook example of an attribution that holds.

## It is also misfiled by the repo's own criterion

`BUILTIN_MITIGATIONS`' header, and `registry/README.md`'s Path 1 / Path 2 split, say
built-ins are for env vars read by a runtime or library, and that anything which only
takes effect because a workload's training script calls `os.environ.get(...)` is a
plugin candidate registered through the `aorta.mitigations` entry-point group.
`DISABLE_TF32` is in the builtins purely on the strength of the hipBLASLt claim. Once
that claim goes, the entry fails its own admission test: the only way it can do
anything is if some workload reads it, which is the definition of Path 2.

## Proposed remedies

**1. The registry entry.** Three options, in my order of preference:

- **Remove it.** It has no runtime consumer, so removal loses no behaviour. This is
  a breaking change for any recipe naming `tf32_off`, but those recipes are currently
  running a duplicate baseline, so the breakage surfaces a bug rather than causing one.
- **Give it a body.** If aorta wants a TF32 axis on ROCm, the mitigation has to drive
  `torch.backends.cuda.matmul.fp32_precision`, which is in-process and therefore a
  plugin against a specific workload, not an env-var bundle. This is the option that
  preserves the intent.
- **Keep it, documented as workload-dependent.** If some private image does read
  `DISABLE_TF32`, that is a Path 2 plugin by this repo's own rule; the entry should
  move there and say which workload honours it. Leaving it in the builtins with the
  hipBLASLt comment is the one option I would argue against.

Whichever is chosen, `env_knobs.py:183` should be corrected or dropped alongside, and
`registry/README.md:34` needs a different qualifying example — `HSA_XNACK` and
`CUDA_LAUNCH_BLOCKING` are already in that table and both genuinely qualify.

**2. The worked example.** Replace `tf32_off` in the probe docs with a mitigation that
has been *measured* to change a verdict. `pytorch_no_cuda_memory_caching`
(`PYTORCH_NO_CUDA_MEMORY_CACHING=1`) is now a good candidate: in an archived 8-cell ×
4-trial probe matrix on gfx950 it is the only registered mitigation of 21 that resolves
a real NaN, with seven cells failing 4/4 under `tier4:nan_signature` and the
`pytorch_no_cuda_memory_caching-none` cell passing 4/4. It has the property the current
example only appears to have: a reader who copies it gets a matrix where the two cells
can actually differ.

**3. A guard, optional but cheap.** Most of this class of bug is detectable without
hardware. A test that asserts every `BUILTIN_MITIGATIONS` env-var name appears
somewhere in the runtime it is attributed to would have caught this at authoring time,
and would catch the next one — upstream removals like `HIPBLASLT_ALLOW_TF32` in
September 2025 turn live mitigations into dead ones silently.

## Reproducing

The matmul comparison needs one GPU and about a minute; the string sweep needs no GPU
at all and is the stronger evidence:

```
docker run --rm rocm/primus:v26.4 bash -lc 'grep -ral "DISABLE_TF32" /opt /usr; \
  echo "--- control ---"; grep -ral "TORCH_ALLOW_TF32_CUBLAS_OVERRIDE" /opt /usr'
```

Verified: the first prints nothing at all; the control prints
`torch/lib/libtorch_cpu.so` and `torch/include/ATen/Context.h`.

## One issue or two?

Filing this as one. The documentation half is not safely separable: a replacement
worked example cannot be chosen without deciding what happens to the registry entry,
and fixing the docs alone would leave a dead built-in that gets adopted again. If the
registry decision needs an owner and stalls, the docs half is a fast follow-up PR that
can land on its own.

**Comment by vivekkhandelwal1** [comment]
Two more registered mitigations that cannot do anything on the stack we ship on, found while building probe scenarios by inverting the registry. Adding them here rather than as separate issues because they are the same pattern this issue describes — but they differ from `tf32_off` in a way that changes the fix, so that distinction is the main thing below.

Stack for everything that follows: `rocm/primus:v26.3`, torch `2.10.0+git94c6e04`, HIP `7.2.53211`, ROCm 7.2.1 in-container, `gfx950:sramecc+:xnack-` (MI350-class), Slurm `meta64`.

## 1. `pytorch_alloc_expandable_segments` — the allocator refuses the option

`PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True`, on the first allocation:

```
UserWarning: expandable_segments not supported on this platform
(Triggered internally at /workspace/pytorch/c10/hip/HIPAllocatorConfig.h:40.)
```

The option is parsed, rejected, and the allocator carries on with its default behaviour, so nothing downstream can depend on it. A fragmentation probe (72 × 64 MiB blocks against a capped pool, every other block freed, then one contiguous request for half the total) reports identical `memory_reserved` (4608 MiB) and `memory_allocated` (2304 MiB) with the option set and unset.

This one has a direct cost to us: `pytorch_alloc_expandable_segments` is a cell in five archived probe matrices, where it is a second baseline under a different name — the exact failure the probe harness exists to prevent.

## 2. `fa_prefer_ck` — the backend is not in the build

`TORCH_ROCM_FA_PREFER_CK=1`, on the first `scaled_dot_product_attention` call:

```
UserWarning: Cannot set preferred SDPA backend to CK since following conditions are not true:
architecture supported for CK: 1, PyTorch built with CK SDPA support: 0
(Triggered internally at /workspace/pytorch/aten/src/ATen/Context.cpp:564.)
```

Note which half fails. **The architecture is supported** — `architecture supported for CK: 1` — and the build is what is missing. fp16 SDPA at `head_dim` 64, 128, 256 and 512 succeeds and produces finite output identically with the variable set to `1`, set to `0`, and unset.

Worth recording alongside it: `fa_prefer_aotriton` sets the same variable to `0`, which is the default, so that entry is a no-op even on a build that *does* have CK.

## Why these are a different kind of dead from `tf32_off`, and why that matters

`tf32_off` is dead because **nobody reads the variable**. `DISABLE_TF32` does not appear anywhere in the image, and the registry's attribution of it to hipBLASLt does not hold; the fix is to correct or remove the entry, and `env_knobs.py:183` already disagrees with `mitigations.py:37` about who consumes it.

These two are the opposite shape. **The variable is read.** PyTorch's allocator config parser reads one and `ATen/Context.cpp` reads the other, each one takes the value, evaluates it, and warns that it cannot comply. The registry entries are *correct* — both would work on a build carrying the capability — so deleting them would be wrong, and so would re-attributing them.

That splits the fix in two:

- `tf32_off` needs its **attribution** corrected, which is a registry-content question.
- These two need a **capability probe**, which is a registry-mechanism question: something that asks, on the stack actually in front of it, whether a mitigation can do anything, and reports a cell as inert rather than letting it run as a silent second baseline. That would also cover a third cause we hit in the same work — `hsa_no_scratch_reclaim` is inert on this image for neither of the above reasons, but because `rocm/primus:v26.3` bakes `HSA_NO_SCRATCH_RECLAIM=1` into `Config.Env`, so the mitigation sets what is already set.

Three causes, one observable: a probe cell that claims to be testing a mitigation and is testing the baseline. The attribution fix does not catch the other two, which is the argument for treating the mechanism half as its own piece of work.

## What would make this checkable

Both warnings above are emitted by the runtime at the moment the knob is evaluated, and both name the reason. A probe that sets a mitigation, runs a trivial workload, and greps the child's stderr for a refusal would catch both of these and cost seconds — it would not catch `tf32_off` (silence is indistinguishable from success there) or `hsa_no_scratch_reclaim` (which needs the image's `Config.Env` compared against the cell's intent). All three are cheap; none of them is the same check.

**Comment by vivekkhandelwal1** [comment]
Filed [#511](https://github.com/ROCm/aorta/issues/511), which generalises this: `tf32_off` turns out to be one of **seven** built-in mitigations that cannot act on the stack probes run on, in four distinct modes — read by nobody (this one, plus `rccl_gfx942_cheap_fence_off`), read and refused (`pytorch_alloc_expandable_segments`, `fa_prefer_ck` — both already reported above), setting a value that is already the default (`fa_prefer_aotriton`, `hsa_enable_cache`), and baked into the image (`hsa_no_scratch_reclaim`).

A single-mitigation issue is the wrong shape for that, so #511 probably supersedes this one — either close this into it, or narrow this to the documentation half, which is still specific to `tf32_off` and still needs a replacement worked example in `registry/README.md`.

Two things #511 adds that are worth flagging here. The attribution conflict recorded above is not isolated: of the 18 distinct variables `BUILTIN_MITIGATIONS` uses, 11 do not appear in `ENV_KNOB_REGISTRY` at all and the other 7 are all marked `INHERITED_UNAUDITED`, so no mitigation variable currently has an audited attribution. And `scripts/audit_env_knobs.py` already performs this exact audit under `--strict` in `gpu-tests.yml` — it is simply pointed at `ENV_KNOB_REGISTRY` and the GEMM libraries rather than at the mitigation registry and the runtime ones.
